File: grpc-java-examples/src/main/java/com/example/grpc/deadline/DeadlineServerStreamingServer.java
package com.example.grpc.deadline;

import com.example.grpc.Delays;
import com.example.grpc.EchoRequest;
import com.example.grpc.EchoResponse;
import com.example.grpc.EchoServiceGrpc;
import com.example.grpc.Servers;
import io.grpc.Context;
import io.grpc.stub.StreamObserver;

import java.io.IOException;
import java.util.logging.Level;
import java.util.logging.Logger;

public class DeadlineServerStreamingServer {

    private static final Logger logger = Logger.getLogger(DeadlineServerStreamingServer.class.getName());

    public static void main(String[] args) throws IOException, InterruptedException {
        Servers.start(new EchoServiceImpl());
    }

    private static class EchoServiceImpl extends EchoServiceGrpc.EchoServiceImplBase {
        @Override
        public void serverStreamingEcho(EchoRequest request, StreamObserver<EchoResponse> responseObserver) {
            logger.log(Level.INFO, "request: {0}", request.getMessage());
            var context = Context.current();

            for (var i = 0; i <= 9; i++) {
                if (context.isCancelled()) {
                    logger.log(Level.WARNING, "cancelled by client: ", context.cancellationCause());
                    return;
                }

                Delays.sleep(i);

                var response = EchoResponse.newBuilder().setMessage("hello " + request.getMessage() + " " + i).build();
                logger.log(Level.INFO, "response: {0}", response.getMessage());
                responseObserver.onNext(response);
            }
            responseObserver.onCompleted();
        }
    }
}

File: grpc-kotlin-examples/src/main/kotlin/com/example/grpc/deadline/DeadlineServerStreamingServer.kt
package com.example.grpc.deadline

import com.example.grpc.EchoRequest
import com.example.grpc.EchoResponse
import com.example.grpc.EchoServiceGrpcKt
import com.example.grpc.Servers
import com.example.grpc.echoResponse
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.onCompletion
import java.util.logging.Logger

object DeadlineServerStreamingServer {
  private val logger = Logger.getLogger(DeadlineServerStreamingServer::class.java.name)

  @JvmStatic
  fun main(args: Array<String>) {
    Servers.start(EchoServiceImpl())
  }

  private class EchoServiceImpl : EchoServiceGrpcKt.EchoServiceCoroutineImplBase() {
    override fun serverStreamingEcho(request: EchoRequest): Flow<EchoResponse> {
      val name = request.message
      logger.info("request: $name")

      return flow {
        for (i in 0..9) {
          val response = echoResponse { message = "hello $name $i" }
          logger.info("response: ${response.message}")
          emit(response)

          delay(i * 1000L)
        }
      }.onCompletion { cause ->
        if (cause != null) {
          logger.warning("stream cancelled or deadline exceeded: ${cause.message}")
        } else {
          logger.info("stream completed successfully")
        }
      }
    }
  }
}
